optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/quad/QuadConstraintStream.java

```diff
@@ -624,4 +625,184 @@ default Constraint rewardConfigurableBigDecimal(String constraintName,
+     * @param constraintWeight never null
+     * @param matchWeigher never null
+     * @return never null
+     */
+    Constraint impactBigDecimal(String constraintPackage, String constraintName, Score<?> constraintWeight,
+            QuadFunction<A, B, C, D, BigDecimal> matchWeigher);
+
+    /**
+     * Positively or negatively impact the {@link Score} by the {@link ConstraintWeight} for each match.
+     * <p>
+     * The constraintWeight comes from an {@link ConstraintWeight} annotated member on the
+     * {@link ConstraintConfiguration}, so end users can change the constraint weights dynamically.
+     * This constraint may be deactivated if the {@link ConstraintWeight} is zero.
+     * If there is no {@link ConstraintConfiguration}, use {@link #impact(String, Score)} instead.
+     * <p>
+     * The {@link Constraint#getConstraintPackage()} defaults to {@link ConstraintConfiguration#constraintPackage()}.
+     * <p>
+     * For non-int {@link Score} types use {@link #impactConfigurableLong(String, ToLongQuadFunction)} or
+     * {@link #impactConfigurableBigDecimal(String, QuadFunction)} instead.
+     * <p>
+     * Use {@code penalizeConfigurable(...)} or {@code rewardConfigurable(...)} unless you intend to mix positive and
+     * negative weights.
+     * @param constraintName never null, shows up in {@link ConstraintMatchTotal} during score justification
+     * @param matchWeigher never null, the result of this function (matchWeight) is multiplied by the constraintWeight
+     * @return never null
+     */
+    default Constraint impactConfigurable(String constraintName, ToIntQuadFunction<A, B, C, D> matchWeigher) {
+        return impactConfigurable(getConstraintFactory().getDefaultConstraintPackage(), constraintName, matchWeigher);
+    }
+
+    /**
+     * As defined by {@link #impactConfigurable(String, ToIntQuadFunction)}.
+     * @param constraintPackage never null
+     * @param constraintName never null
+     * @param matchWeigher never null
+     * @return never null
+     */
+    Constraint impactConfigurable(String constraintPackage, String constraintName,
+            ToIntQuadFunction<A, B, C, D> matchWeigher);
+
+    /**
+     * Positively or negatively impact the {@link Score} by the {@link ConstraintWeight} for each match.
+     * <p>
+     * The constraintWeight comes from an {@link ConstraintWeight} annotated member on the
+     * {@link ConstraintConfiguration}, so end users can change the constraint weights dynamically.
+     * This constraint may be deactivated if the {@link ConstraintWeight} is zero.
+     * If there is no {@link ConstraintConfiguration}, use {@link #impact(String, Score)} instead.
+     * <p>
+     * The {@link Constraint#getConstraintPackage()} defaults to {@link ConstraintConfiguration#constraintPackage()}.
+     * <p>
+     * Use {@code penalizeConfigurableLong(...)} or {@code rewardConfigurableLong(...)} unless you intend to mix
+     * positive and negative weights.
```

Comment:
Soft suggestion:
I'd use the word "instead" in these sentences for clarity.
Also "you intend to mix positive and negative weights" could be interpreted in general, not just for this constraint. We really want to discourage them using "impact".

Suggestion:
 "... unless you intend to mix positive and negative weights"
 -> "... instead, unless this constraint can both have positive and negative weights."

Also, I 'd always put it as the second paragraph, as I think this _nudging_ (yes that's what it is :)), is more important what getConstraintPackage() etc default to.